Kernel+LibKeyboard+KeyboardMapper+keymap: All things keyboard!

BenWiederhake opened the pull request:
I tried to vaguely group the commits, but they're still pretty connected with each other.
- Kernel: new `getkeymap` syscall
- KeyboardMapper: Fix GUI and logic errors, start with current keymap by default (if no args given)
- keymap(1): fix unveil errors, better error reporting, support querying
- LibKeyboard: return error instead of asserting, support querying, don't force every application to carry the default keymap (which weighs 2.5 KiB)
- Actual keymaps: Fix some errors, extend `de.json` by a lot
- Meta: Add a script that checks the keymaps for consistency. Note that `jp.son` defines some exotic keycodes!
- Documentation :)

linusg commented on Documentation/BuildInstructions.md:
```diff
@@ -169,3 +169,17 @@ Later on, when you `git pull` to get the latest changes, there's (usually) no ne
+
+set -e
+
+echo $'\n[keymap]\nExecutable=/bin/keymap\nArguments=de\nUser=anon' >> mnt/etc/SystemServer.ini
```
If you want increased readability:

```suggestion
cat <<EOT >> mnt/etc/SystemServer.ini

[keymap]
Executable=/bin/keymap
Arguments=de
User=anon
EOT
```

(this is what I use locally)

emanuele6 replied:
you should use `<< 'EOT'` (with single quotes) to prevent variable substitution.

BenWiederhake replied:
Ah, yes! I like one-liners too much ^^'

linusg commented on Documentation/BuildInstructions.md:
````diff
@@ -169,3 +169,17 @@ Later on, when you `git pull` to get the latest changes, there's (usually) no ne
+echo $'\n[keymap]\nExecutable=/bin/keymap\nArguments=de\nUser=anon' >> mnt/etc/SystemServer.ini
+```
+
+This will configure your keymap to German (`de`) instead of English. See `Base/res/keymaps/` for a full list.
````
Let's make it easier for the reader and link this:

```suggestion
This will configure your keymap to German (`de`) instead of English. See [`Base/res/keymaps/`](../Base/res/keymaps/) for a full list.
```

linusg commented on Documentation/BuildInstructions.md:
````diff
@@ -169,3 +169,17 @@ Later on, when you `git pull` to get the latest changes, there's (usually) no ne
+Create a file with the exact name `sync-local.sh` in the project root (the same directory as `.clang-format`), with content like this:
+
+```
+#!/bin/bash
````
This is misleading as the interpreter is explicitly given here and the shebang therefore ignored:

https://github.com/SerenityOS/serenity/blob/ec57c6c83601d354e3103e4c668c3ef4a4cadf8f/Meta/build-root-filesystem.sh#L145

(`sh` may or may not be bash, we could make it say bash explicitly or just rely on the shebang)

linusg commented on Meta/lint-keymaps.py:
```diff
@@ -0,0 +1,126 @@
+    all_good = True
+
+    if not isinstance(values, list):
+        report(filename, '"{}" is not a list'.format(mapname))
```
Let's use JSON terminology here:

```suggestion
        report(filename, '"{}" is not an array'.format(mapname))
```

linusg commented on Meta/lint-keymaps.py:
```diff
@@ -0,0 +1,126 @@
+
+    for i, c in enumerate(values):
+        if len(c) > 1:
+            report(filename, 'more than one character (“{}”) for charmap index {} of {}'.format(c, i, mapname))
```
Can we just use "normal" quotes here?

```suggestion
            report(filename, 'more than one character ("{}") for charmap index {} of {}'.format(c, i, mapname))
```

BenWiederhake replied:
But but but … *fancy*!

Yes, absolutely, I changed it :)

linusg commented on Meta/lint-keymaps.py:
```diff
@@ -0,0 +1,126 @@
+    all_good = True
+
+    if not isinstance(fullmap, dict):
+        report(filename, 'is not a dict')
```
Same as above,

```suggestion
        report(filename, 'is not an object')
```

linusg reviewed (commented):
\\o/ for keymap linting!

alimpfard commented:
I was gonna ask, but why not stick a `keymap de` in your shellrc? the terminal runs at boot anyway?
the current way is definitely less hacky though~

bcoles commented on Userland/Utilities/keymap.cpp:
```diff
@@ -42,20 +42,45 @@ int main(int argc, char** argv)
+    auto character_map = Keyboard::CharacterMap::load_from_file(path);
+    if (!character_map.has_value()) {
+        warnln("Cannot read keymap {}", path);
+        warnln("Hint: Must bei either a keymap name (e.g. 'en') or a full path.");
```
```suggestion
        warnln("Hint: Must be either a keymap name (e.g. 'en') or a full path.");
```

emanuele6 commented:
I noticed that we are not being consistent with the style of keymap files.
Some of them only write non-ascii characters as backslash sequences (`it.json`, `fr.json`, ...) and some others write the non-ascii characters literally (`gb.json`, `de.json`, ...).

We should just pick one style and be consistent.

BenWiederhake commented:
> I was gonna ask, but why not stick a `keymap de` in your shellrc? the terminal runs at boot anyway?
> the current way is definitely less hacky though~

Hmm, we can't change it in `Base` because it's in version control. So we would need a command `sync-local.sh` that adds it to `.shellrc`. But that would be strictly worse than making the call *once* from SystemServer, because `.shellrc` will be executed multiple times, whenever the user opens a Terminal/Shell/Hack Studio/etc.

So for now, I'd like to stay with the SystemServer command.

Alternatively, we go the route that many system-utilities do: Make "the configuration file" a folder that gets executed alphabetically. This way we could tell git to ignore `Base/etc/SystemServer/*-local`, and the user only needs to create the file once. However, this is out of scope for this already-too-large-PR.

BenWiederhake commented:
> I noticed that we are not being consistent with the style of keymap files.
> Some of them only write non-ascii characters as backslash sequences (`it.json`, `fr.json`, ...) and some others write the non-ascii characters literally (`gb.json`, `de.json`, ...).
> 
> We should just pick one style and be consistent.

I agree! I suggest inserting the actual utf-8 things: It's easier to read, and for users who usually type it on their own keyboard, it's easier to write, too.

In that case, we would need to change `be`, `fr`, `it`, `trq`. Since both the lint script and `CharacterMapFile::load_from_file` do not see any difference, I'm too lazy to touch it, though.

BenWiederhake commented:
And because I forgot it when I created this PR: Please enjoy this tasteful screenshot.

![Bildschirmfoto_2021-01-30_21-59-26](https://user-images.githubusercontent.com/2690845/106381250-b4e4f180-63b7-11eb-9590-7494047687d7.png)

And in case anybody is interested: `KeyboardMapper` should also allow entering the Unicode codepoint somehow, because otherwise the only way to enter a "completely new" codepoint is to copy-paste it from somewhere else.

BenWiederhake commented:
Changes since first push:
- Fix silly python lint error
- Adjusted terminology in script
- More readable documentation
- Typos in the code and commit messages
- Rebased for convenience

BenWiederhake commented:
Changes since last push:
- Link in the documentation as linus suggested

emanuele6 commented:
> I'm too lazy to touch it, though.

@BenWiederhake, you don't have to do it by hand~ :laughing:

```bash
#!/bin/bash

< <(grep -hoP -e '\\u[A-Za-z0-9]{4}' ./*.json) grep -i -ve '\\u001b' \
| sort -u \
| while read -r; do
    sed -i "s,\\$REPLY,$(eval "echo $'$REPLY'"),g" ./*.json
done
```
![file](https://user-images.githubusercontent.com/20175435/106381951-80276900-63bc-11eb-8e54-e6a7fae9bd38.png)

BenWiederhake commented:
Changes since last push:
- Unicode replacement through arcane incantation incantation (thanks @emanuele6!)
- Rebased for convenience

BenWiederhake commented:
Changes since last push:
- Fix	tabs	and	spaces	in	keymaps	by	just	spaces
- Rebased yet again

awesomekling commented on Kernel/Syscalls/keymap.cpp:
```diff
@@ -66,4 +67,35 @@ int Process::sys$setkeymap(Userspace<const Syscall::SC_setkeymap_params*> user_p
 
+int Process::sys$getkeymap(Userspace<const Syscall::SC_getkeymap_params*> user_params)
+{
+    // FIXME: Which pledge is appropriate here?
```
`stdio` seems fine

awesomekling replied:
But a new `getkeymap` might be good too, since it's kind of a niche thing(?)

awesomekling commented on Userland/Utilities/keymap.cpp:
```diff
@@ -49,13 +49,26 @@ int main(int argc, char** argv)
+    dbgln("path is at {:p}, and contains {}", path, path);
+
+    if (!path) {
+        auto keymap = Keyboard::CharacterMap::fetch_system_map();
+        if (!keymap.has_value()) {
+            perror("getkeymap");
```
Please don't use `errno` to carry errors outside `LibC`. Instead, return a `Result<T, OSError>` or similar.

linusg commented on Documentation/BuildInstructions.md:
````diff
@@ -169,3 +169,23 @@ Later on, when you `git pull` to get the latest changes, there's (usually) no ne
+EOF
+```
+
+This will configure your keymap to German (`de`) instead of English. See [`Base/res/keymaps/`](../Base/res/keymaps/) for a full list.
````
```suggestion
This will configure your keymap to German (`de`) instead of US English. See [`Base/res/keymaps/`](../Base/res/keymaps/) for a full list.
```

emanuele6 replied:
we should also rename `en.json` to `us.json` or `en.json` to `enus.json` and `gb.json` to `engb.json` (for consistency with `ptpt` and `ptbr`), no?

linusg replied:
Sounds good, yeah. I'd prefer `en-us`, `en-gb`, `pt-...` though.

emanuele6 replied:
agreed

BenWiederhake replied:
Okay, this is *definitely* out of scope for this PR. I mainly wanted to get the German layout complete and all tools working properly.

(Oh and: language-country codes are usually written like `pt_BR`, aren't they?)

emanuele6 replied:
@BenWiederhake: this is the full output of `locate keymaps` on my computer: https://gist.github.com/emanuele6/d89f513943fa26d67f0a8d1f6c6c294c

`pt-br` seems to be how is usually written. `pt_BR` looks like the name format of `locale` files.

linusg commented on Meta/lint-keymaps.py:
```diff
@@ -0,0 +1,128 @@
+
+
+if __name__ == '__main__':
+    os.chdir(os.path.dirname(__file__) + "/../Base/res/keymaps/")
```
No change required as far as I'm concerned, this is just a linting script... but you should try [`pathlib`](https://docs.python.org/3/library/pathlib.html) if you haven't yet :)

BenWiederhake replied:
It doesn't really help with going from `__file__` to the containing directory – at best it's another function. Simpler than addition is impossible (although I really like their `/` operator!), and `chdir` is a OS-level function, I don't really like to see it wrapped. So I'm sure `pathlib` is nice, but it solves a different problem :)

linusg replied:
You shouldn't chdir in the first place, listdir (and the pathlib equivalent) works with explicitly given paths as well. Not here to argue though, if it works, it works

BenWiederhake commented:
Changes since last push:
- Call the default keyboard "US English"
- LibKeyboard no longer touches `errno`
- Remove resolved `//FIXME` (getkeymap(2) pledge)

BenWiederhake commented:
Changes since last push:
- Converted `sys$getkeymap` to be new pledge (new feature!)
- Added the pledge to keymap(1) and KeyboardMapper
- Convenience for rebased :P
